### DISCUSSION TITLE
What are the rules behind "You have triggered an abuse detection mechanism"?

### DISCUSSION THREAD

**Discussion by payne911** [community/community#24307] [Apps, API and Webhooks]
<p>Hi,</p>
<p>I am developing <a href="https://github.com/useful-forks/useful-forks.github.io" rel="noopener nofollow ugc">a tool</a> which helps increase the discoverability of useful forks on GitHub.</p>
<p>Despite providing an Access Token to authenticate my requests to the GitHub API, I get a <code>403</code> error message mentioning an “abuse detection mechanism” after a bit <strong>less than 500 requests</strong> (477, to be more precise).</p>
<p>Unfortunately for me, it seems like I have reached the limit of the documentation: there is almost nothing to help people figure out what they should do when they see that error, <em>nor what causes it exactly</em>.</p>
<hr>
<p>I am currently using REST calls to the following endpoints:</p>
<ul>
<li>
<strong>GET repo</strong> : <code>https://api.github.com/repos/${user}/${repo}</code>
</li>
<li>
<strong>LIST forks</strong> : <code>https://api.github.com/repos/${user}/${repo}/forks?sort=stargazers&amp;per_page=${FORKS_PER_PAGE}&amp;page=${page_number}</code>
</li>
<li>
<strong>Compare two commits</strong> : <code>https://api.github.com/repos/${user}/${repo}/compare/${parentDefaultBranch}...${extract_username_from_fork(currFork.full_name)}:${currFork.default_branch}</code>
</li>
</ul>
<p>The idea is to find all the forks of a queried repository, and then list only the ones which have seen activity since they were created (because the large majority of forks never ever see any activity and simply pollute that list). The “compare two commits” is only there to see if the fork is “<code>ahead_by</code>” at least 1 commit.</p>
<p>At this point, it is very much unclear what I should do to maximize the amount of forks I can scan:</p>
<ol>
<li>Is there a way to apply for a “GitHub integration developer program” of some sort that would lend my project a special token to use ?</li>
<li>How much time should I wait between each request ?</li>
<li>Do the “<code>OPTIONS</code>” requests automatically sent before the “<code>GET</code>” requests count toward the “maliciousness counter” ? Should I try to make it so my browser avoids making the “<code>OPTIONS</code>” requests ?</li>
<li>Would swapping my implementation to use GraphQL make things any better ? How to know in advance if I would not be spending hours swapping to GraphQL just to realize that somehow my requests would then be restricted even sooner ?</li>
<li>
<s>The “<code>retry-after: 60</code>” header which is part of the <code>403</code> response : is it in seconds, or minutes ?</s> Answer: It’s seconds.</li>
</ol>
<p>Thanks ! :slight_smile:</p>

**Comment by payne911** [comment]
<p>I just realized <a href="https://docs.github.com/en/rest/overview/resources-in-the-rest-api#abuse-rate-limits" rel="noopener nofollow ugc">the documentation page</a> could be made to point to <a href="https://docs.github.com/en/rest/guides/best-practices-for-integrators#dealing-with-abuse-rate-limits" rel="noopener nofollow ugc">the right section</a> instead of sending you at the top of that other page.</p>
<p>For others like me which hadn’t read the whole page once they clicked on the link, there <em>is</em> further information there:</p><aside class="onebox githubfolder">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">
      <a href="https://docs.github.com/en/rest/guides/best-practices-for-integrators#dealing-with-abuse-rate-limits" target="_blank" rel="noopener nofollow ugc">docs.github.com</a>
  </header>
  <article class="onebox-body">
    

<h3><a href="https://docs.github.com/en/rest/guides/best-practices-for-integrators#dealing-with-abuse-rate-limits" target="_blank" rel="noopener nofollow ugc">Best practices for integrators - GitHub Docs</a></h3>

<p><a href="https://docs.github.com/en/rest/guides/best-practices-for-integrators#dealing-with-abuse-rate-limits" target="_blank" rel="noopener nofollow ugc">//docs.github.com/en/rest/guides/best-practices-for-integrators</a></p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>(I’m just leaving this here for posterity, for I’d still love to hear answers to my questions in the OP.)</p>

**Comment by payne911** [comment]
<p>Moreover, again for posterity, some good discussions (and implementations!):</p>
<ul>
<li><a href="https://github.com/octokit/rest.js/issues/526" class="inline-onebox" rel="noopener nofollow ugc">Implement recommended best practises to prevent abuse rate limits · Issue #526 · octokit/rest.js · GitHub</a></li>
<li><a href="https://github.com/octokit/rest.js/issues/1099" class="inline-onebox" rel="noopener nofollow ugc">🔌 Request throttling plugin · Issue #1099 · octokit/rest.js · GitHub</a></li>
</ul>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="payne911" data-post="1" data-topic="164457">
<div class="title">
<div class="quote-controls"></div>
 payne911:</div>
<blockquote>
<p>The idea is to find all the forks of a queried repository, and then list only the ones which have seen activity since they were created (because the large majority of forks never ever see any activity and simply pollute that list).</p>
</blockquote>
</aside>
<p>I’m not sure if comparing the default branches is the best way to do that comparison. I don’t have statistics on how people work, but for changes I intend to contribute I’ll usually commit them to a feature branch. That’s also what the Github web editor encourages.</p>

**Comment by payne911** [comment]
<aside class="quote no-group" data-username="airtower-luna" data-post="4" data-topic="164457">
<div class="title">
<div class="quote-controls"></div>
 airtower-luna:</div>
<blockquote>
<p>I’m not sure if comparing the default branches is the best way to do that comparison. I don’t have statistics on how people work, but for changes I intend to contribute I’ll usually commit them to a feature branch. That’s also what the Github web editor encourages.</p>
</blockquote>
</aside>
<p>Many considerations come into play for choosing that criteria:</p>
<ol>
<li>The goal is to find separately-active forks. Forking to commit to a feature-branch and then making a PR to merge the work back into the parent : that is not a relevant fork as the work will be back in the parent at some point in time, or at least be visible as a PR in the project. In fact, I would argue those forks mostly just pollute the tree (hence this project which filters them out).</li>
<li>I am already hitting different limitations which relate to the amount of API calls I’m making: while there are very many other criterias I wish I could add, they would hinder the depth of the scan in a significant manner. If I had no Rate-Limit, then I would love to go crazy on that. :slight_smile:
</li>
</ol>
<p>This tool’s goal really is to find forks which might have picked up from a dead open-source project, or which are trying to diverge from the original work and become their own project.</p>

**Comment by airtower-luna** [comment]
<p>Unmerged feature branches or pull requests might still have useful fixes or features, but yeah, the rate limit is hard to argue with. :sweat_smile:</p>

**Comment by payne911** [comment, marked as answer]
<p>For anyone interested, here is the PR which solved my problems: <a href="https://github.com/useful-forks/useful-forks.github.io/pull/15" class="inline-onebox" rel="noopener nofollow ugc">Octokit integration by payne911 · Pull Request #15 · useful-forks/useful-forks.github.io · GitHub</a></p>
